src/components/widgets/Toggle.tsx

```diff
@@ -0,0 +1,247 @@
+  themeRole?: ThemeRole;
+  /** Size of toggle. Optional, defaults to medium. */
+  size?: "medium" | "small";
+  /** Specification on how toggle should be styled. Optional. */
+  styleOverrides?: ToggleStyleSpecSubset;
+};
+
+/** Fully controlled Toggle component. */
+export default function Toggle({
+  value,
+  onToggle,
+  disabled,
+  label,
+  labelPosition = "left",
+  ariaLabel,
+  themeRole,
+  size = "medium",
+  styleOverrides,
+}: ToggleProps) {
+  const theme = useUITheme();
+  const [hoverState, setHoverState] = useState<"default" | "hover">("default");
+
+  const styleSpec: ToggleStyleSpec = useMemo(() => {
+    const themeColor = theme && themeRole && theme.palette[themeRole];
+    const mainColor = themeColor?.hue ?? blue;
+    const mainLevel = themeColor?.level ?? 400;
+    const labelColor = gray[600];
+    const disabledColor = gray[300];
+
+    const defaultStyleSpec: ToggleStyleSpec = {
+      container: {},
+      default: {
+        off: {
+          backgroundColor: "none",
+          knobColor: mainColor[mainLevel],
+          borderColor: mainColor[mainLevel],
+          labelColor,
+        },
+        on: {
+          backgroundColor: mainColor[mainLevel],
+          knobColor: "white",
+          borderColor: mainColor[mainLevel],
+          labelColor,
+        },
+      },
+      hover: {
+        off: {
+          backgroundColor: mainColor[100],
+          knobColor: mainColor[mainLevel],
+          borderColor: mainColor[mainLevel],
+          labelColor,
+        },
+        on: {
+          backgroundColor: mainColor[Math.min(mainLevel + 200, 900)],
+          knobColor: "white",
+          borderColor: mainColor[Math.min(mainLevel + 200, 900)],
+          labelColor,
+        },
+      },
+      disabled: {
+        off: {
+          backgroundColor: "none",
+          knobColor: disabledColor,
+          borderColor: disabledColor,
+          labelColor,
+        },
+        on: {
+          backgroundColor: disabledColor,
+          knobColor: "white",
+          borderColor: disabledColor,
+          labelColor,
+        },
+      },
+    };
+
+    return merge({}, defaultStyleSpec, styleOverrides);
+  }, [styleOverrides, theme, themeRole]);
+
+  /**
+   * The CSS styles that should be applied can depend
+   * on (1) whether or not the component is
+   * focused/hovered, (2) which option is selected,
+   * and (3) whether the toggle is disabled.
+   */
+  const currentStyles = useMemo(() => {
+    const selectedOption = value ? "on" : "off";
+
+    return disabled
+      ? styleSpec.disabled[selectedOption]
+      : styleSpec[hoverState][selectedOption];
+  }, [hoverState, value, styleSpec, disabled]);
+
+  const finalAriaLabel =
+    ariaLabel ?? (typeof label === "string" ? undefined : "Toggle");
+  const labelId = useMemo(
+    () =>
+      ariaLabel === undefined && typeof label === "string"
+        ? uniqueId("toggle-label-")
+        : undefined,
+    [label, ariaLabel]
+  );
```

Comment:
I **_think_** the logic should be more like: 
- if we have a label at all, we use `aria-labelledby` and `aria-label` is undefined
- if no label is passed in, we use the `ariaLabel` prop or fall back to "toggle"

I played around with VoiceOver some and it does seem that if something ends up with both `aria-labelledby` and `aria-label`, VoiceOver ignores the `aria-label` and instead reads the `aria-labelledby`. I also tested some nested labels and it seemed to read the nested text content fine. Not sure how other screenreaders handle either situation though.

Reply:
I'm fine with this logic, except for that I think we should prefer the Toggle's `ariaLabel` over the `label` prop so that the user can explicitly provide the aria label if desired. The use may provide a `label` React node that has no text at all, for instance, so giving an explicit `ariaLabel` would be better in that case.

Reply:
> If the label does not include descriptive text, then you should pass in an ariaLabel prop to override

Here the 'you' is the user of this component, yes? So we should add these directions to the comments.

Reply:
Yes, that was my intention to simply bolster the props' documentation so that consumers of the API understand the nuances and what to expect when passing or not passing these two props. I guess a better way to document the `label` prop is along these lines:
> If a label is provided, then aria-labelledby will be used to connect the label's text to the toggle. If the label does not include descriptive text, then an ariaLabel prop should be included to override this behavior.